File: solutions/hard/23.Merge_k_Sorted_Lists/23.Merge_k_Sorted_Lists.py
# Definition for singly-linked list.
from ast import List
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        common_array = []
        k = 0
        while k < len(lists):
            while lists[k]:
                common_array.append(lists[k].val)
                lists[k] = lists[k].next
            k += 1
        print(common_array)
        if len(common_array) == 0:
            return None
        sorted_arr = sorted(common_array)

        head = ListNode(sorted_arr[0])
        tail = head
        for el in range(1, len(sorted_arr)):
            node = ListNode(sorted_arr[el])
            if tail is None:
                tail = head
            tail.next = node
            tail = node

        return head

File: solutions/hard/23.Merge_k_Sorted_Lists/23.Merge_k_Sorted_Lists.ts

//Definition for singly-linked list.
class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.next = (next===undefined ? null : next)
    }
}


function mergeKLists(lists: Array<ListNode | null>): ListNode | null {
  const newArr:number[] = []
  let k = 0
  while(k < lists.length) {
    while(lists[k]){
      newArr.push(lists[k].val)
      lists[k] = lists[k].next
    }
    k++
  }
  if(newArr.length === 0) return null
  const sortedArr = newArr.sort((a, b) => a - b)

  let list = new ListNode(sortedArr[0]);
  let tail = list;
  for (let i = 1; i < sortedArr.length; i++) {
    if (tail.next === null) {
        const node = new ListNode(sortedArr[i]);
        tail.next = node;
        tail = node;
    }
  }

  return list
};